### PR TITLE
mathesar-424 Handle `TIMESTAMP` data type in the backend

### LINKED ISSUE #424: Handle `TIMESTAMP` data type in the backend
This issue is to ensure that Mathesar can handle the Postgres `TIMESTAMP` data type.

As part of this issue, we need to ensure that:
- Users can use the API to change a column to `TIMESTAMP`(with time zone) if it's possible to do so.
- Users can use the API to change a column to `TIMESTAMP`(without time zone) if it's possible to do so.
- Users can set and change the `precision` of a given `TIMESTAMP` column via a `type_options` field in the API.
- Automatic type inference during file import suggests `TIMESTAMP` when it makes sense to do so. 
  - We could use something like https://dateparser.readthedocs.io/

**Additional Context** 
- #250 (parent issue)
- See PR #368 for code that calculates which types are allowed for a column.
- This issue was previously specced out in #128

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #424 Handle `Timestamp` data type in the backend.

This PR introduces Timestamp data type(with and without timezone) and relevant functions to convert text data types that contain a date-time string into a `timestamp` data type.

The order by which the data type is inferred will change to the following
1. If a text contains a date string without any time information `date` type will be inferred.
2. If a text contains a date string along with time information but does not include a `timezone` information `timestamp without timezone` data type will be inferred.
3. If a text contains a date string along with time and timezone information, `timestamp with timezone` data type will be introduced. This type will be inferred if only one  or more rows contains the necessary information as other data type will end up with losing the information 

**Technical details**
Adds TImestamp with timezone and Timestamp without timezone data type and  Relevant casting function that restricts the data that can be converted to the data type.
The casting function of `date` data type has been limited to accept only `text` with `date` element that does not contain `time` information
**Screenshots**

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

## TaskList
 
- [x] Create a timestamp column(db/tests/columns/operations/test_create.py)
- [X] Automatic type inference during file import suggests (with time zone or without timezone) when it makes sense to do so (db/tests/tables/operations/test_infer_types.py:31)
- [x] Users can use the API to change a column to TIMESTAMP(with time zone or without timezone) if it's possible to do so (db/tests/types/operations/test_cast.py:169)
- [x] Users can set and change the precision of a given TIMESTAMP column via a type_options field in the API(db/tests/types/operations/test_cast.py:825)
- [x] Fetching a timestamp record(mathesar/tests/api/test_record_api.py)

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/tests/types/operations/test_cast.py
Comment (silentninja) [inline db/tests/types/operations/test_cast.py:40]: `Date` was repeated, so deleted this duplicate variable

File: db/tests/types/operations/test_cast.py
Comment (dmos62) [inline db/tests/types/operations/test_cast.py:472]: I presume a timestamp should be castable to any text/character type, including `CHAR`.

Comment (silentninja) [reply db/tests/types/operations/test_cast.py:472]: It is possible to cast a `timestamp` to `CHAR`, but `CHAR` returns exact number of characters specified and this causes problems with testing as the `type_options` has to be specified with each test case, which is not possible right now. I have created a [issue](https://github.com/centerofci/mathesar/issues/869) that should help with testing such cases

**Review by kgodey — COMMENTED** [review]
I'll leave this to @mathemancer to review/merge. I took a quick glance and don't see any issues.

**Comment by kgodey** [comment]
Updated status to blocked since it's blocked by #628

File: db/types/operations/cast.py
Comment (mathemancer) [inline db/types/operations/cast.py:626]: I think `TIMESTAMP` types should also be valid sources for dates (in case there's a timestamp value at midnight so the cast would be lossless).

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, pretty good.  I have a few requests, though.

Please see the PostgreSQL docs about `RETURNS NULL ON NULL INPUT`:  https://www.postgresql.org/docs/13/sql-createfunction.html

This will let you take lots of checks for `NULL` out of your SQL function code (and also make things _much_ more efficient).

Also, I think we should be able to cast `TIMESTAMP` types to `DATE` whenever the cast would be lossless.  I'd kind of expect to be able to do such a cast if I had a bunch of values in my column at midnight.

See some in-line comments for other requests.

**Comment by mathemancer** [comment]
@silentninja Have you noticed the requested changes on this PR?

**Comment by silentninja** [comment]
> @silentninja Have you noticed the requested changes on this PR?

Yes, I did notice it, sorry for not changing the status. I got caught up with other issues, will be working on it today.

File: db/types/operations/cast.py
Comment (silentninja) [reply db/types/operations/cast.py:540]: I wanted to have a strict inference and without relying on the order of the datatype in `INFERENCE_DAG` but missed out on this edge case. If we end up removing the comparison with date value, we would also be allowing timestamp without time `May-2007-29`.

**Comment by silentninja** [comment]
> see the PostgreSQL docs about RETURNS NULL ON NULL INPUT: https://www.postgresql.org/docs/13/sql-createfunction.html

I didn't want to modify the `db.types.operations.cast.assemble_function_creation_sql` as it is being used by other casting functions. 

Moreover, I am not sure if this [special case where performance drops](https://stackoverflow.com/questions/8455177/function-executes-faster-without-strict-modifier) applies in our situation as it mostly applies to table functions but wanted to give it the benefit of doubt

File: db/types/operations/cast.py
Comment (silentninja) [reply db/types/operations/cast.py]: Thanks for pointing this out, felt like I was missing something. As for wrapping it in a transaction, I think you mean the query that is calling this function like the `ALTER TABLE`, I might have to mention that the casting function should be in its own transaction block somewhere. Spent quite some time looking for a better alternative but unfortunately had to end up with this.

**Review by mathemancer — APPROVED** [review]
I think this should work, and the choices w.r.t. what casts to what else seem logical.

About setting the functions to `RETURNS NULL ON NULL INPUT`, that should be safe since these are all casting functions that definitely _ought_ to return null on null input.  Also, many of them can't be inlined for various reasons.  I'll make a separate issue to explore that, though, since some of them _can_ be inlined, and these casting functions get run for every row of some table or another quite often.  For now, it's a moot point, since the null checks all got removed (unless we do see some improvement overall in performance).

**Comment by silentninja** [comment]
@mathemancer Thanks! I will change the strict keyword to `RETURNS NULL ON NULL INPUT` as it seems be descriptive compared to `STRICT`

**Comment by silentninja** [comment]
@kgodey We need to specify that the `datetime casting functions` are not transaction-safe and can affect timezones when run in the same transaction, where should this be documented?

**Comment by kgodey** [comment]
@silentninja Please document it in the code and in the wiki (write a short spec that's linked to from "Technical Specs" in https://wiki.mathesar.org/en/engineering/architecture). Please also explain what "can affect timezones" means in greater detail.

**Comment by kgodey** [comment]
@silentninja Feel free to merge this PR when you're ready.

**Pull request merged**
